CI: Implementing benchmark regression tests

ChinYikMing opened the pull request:
Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.

We also want "workflow_dispatch" because we require a method for running the benchmark the first time so that we can store the base benchmark and utilize it for benchmark comparison in the future.

Before merging this PR, please do the following:
1. Create a branch called "gh-pages", this is the default branch for gitHub repository to deploy our web-based  benchmark visualization. This branch can be left only ".git" directory.
2. Go to the setting of  repository  > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.

Last, the benchmark CI should work!

Close #166

jserv commented on .github/workflows/main.yml:
```diff
@@ -55,3 +55,43 @@ jobs:
             .ci/check-format.sh
       shell: bash
+
+  benchmark:
```
Move benchmark part out of `main.yml` since the latter is complex enough.

ChinYikMing commented:
One more thing: you can setup the self-hosted runner for consistency of performance(hopefully) afterwards since I do not have permission to do that.

ChinYikMing committed:
CI: Consolidate benchmark action

Using "pull_request_target" is required since
benchmark action updates data in the "gh-pages" branch
for visualization in gitHub pages, and "pull_request"
does not grant the "GITHUB_TOKEN" needed for that purpose.

We also want "workflow_dispatch" because we require a method
for running the benchmark the first time so that we can store
the base benchmark and utilize it for benchmark comparison
in the future.

Since the benchmark action has already been executed once when
a pull request is opened, we also need to filter out the
merging push event.

Close #166

ChinYikMing commented:
[e829d89](https://github.com/sysprog21/rv32emu/pull/213/commits/e829d89661ff702ad976cf161c3922756876b8ce) changes the following:

1. Since the benchmark action has already been executed once when a pull request is opened, we also need to filter out the merging push event.
2. Run the "changed-files" action in higher order

For further merging the pull requests, please do not modify the commit message "Merge pull request " since it is the conditional of benchmark action not to run again after merging.

jserv commented:
> Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.

Can we refrain from creating a `gh-pages` branch in this repository? I think this repository should remain as minimal as possible, primarily preserving git objects for source code and necessary materials. If needed, we can have a `gh-pages` branch in another repository.

ChinYikMing commented:
> > Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.
> 
> Can we refrain from creating a `gh-pages` branch in this repository? I think this repository should remain as minimal as possible, primarily preserving git objects for source code and necessary materials. If needed, we can have a `gh-pages` branch in another repository.

Sure, a branch can be called whatever in another repository, I think the branch can be called as `rv32emu_benchmark` which makes more sense. 

Please give me URL of the repository that used to store the benchmark data.

jserv commented:
> Please give me URL of the repository that used to store the benchmark data.

Let's try this: https://github.com/sysprog21/rv32emu-bench

ChinYikMing commented:
Please add a secret called `RV32EMU_BENCH_TOKEN`(other naming is OK, if changes to other please inform me) to `rv32emu`  repository, this token should be a PAT that having permission in scope of repo.

ChinYikMing committed:
CI: Consolidate benchmark action

To minimalize the rv32emu repository,
push benchmark action data to a different repository.

ChinYikMing commented:
Before merging this PR, go to the setting of `rv32emu` repository > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.

Now, the benchmark CI should work and benchmark data are stored in `rv32emu-bench` repository.

[Link of benchmark data visualization](https://sysprog21.github.io/rv32emu-bench/)

jserv commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,51 @@
+    if: contains(toJSON(github.event.head_commit.message), 'Merge pull request ') == false
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v3.6.0
```
Replace `v3.6.0` with recent `v4`.

jserv commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,51 @@
+            sudo apt-get update
+            sudo apt-get install libsdl2-dev libsdl2-mixer-dev
+            sudo pip3 install numpy
+            sh .ci/riscv-toolchain-install.sh
```
It is not necessary to run `ci/riscv-toolchain-install.sh`.

jserv commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,51 @@
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' ||
+            github.event_name == 'workflow_dispatch'}}
+        run: |
+          ./tests/benchmark_combiner.py
```
Specify `tests/benchmark_combiner.py` instead.

jserv reviewed (changes requested):
Rename `tests/benchmark_combiner.py` to `tests/bench-aggregator.py`.

ChinYikMing commented:
> Rename `tests/benchmark_combiner.py` to `tests/bench-aggregator.py`.

Done.

jserv commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+            github.event_name == 'workflow_dispatch'}}
+        run: |
+            sudo apt-get update
+            sudo apt-get install libsdl2-dev libsdl2-mixer-dev
```
SDL is not required for building.

jserv commented:
> Please add a secret called `RV32EMU_BENCH_TOKEN` to rv32emu repository, this token should be a PAT that having permission in scope of repo.

DONE.

> go to the setting of rv32emu repository > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

DONE.

jserv commented:
I defer to @qwe661234 for validating the scripts in `tests` directory.

jserv commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,48 @@
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' ||
+            github.event_name == 'workflow_dispatch'}}
+        run: |
+          tests/bench_aggregator.py
```
Name it `bench-aggregator.py` without an underscore.

jserv commented on tests/bench-aggregator.py:
```diff
@@ -0,0 +1,31 @@
+
+# combine benchmarks output data
+benchmarks_output = [
+    "dhrystone_output.txt", 
```
I suppose they are in the form of JSON, right? If so, use `json` as filename suffix.

ChinYikMing replied:
Yes

ChinYikMing committed:
CI: Bump dependency and refactor workflow

qwe661234 commented:
What information we would receive if the benchmark fails?

ChinYikMing commented:
> What information we would receive if the benchmark fails?

Which part of failure in workflow that you referred to? 

Normally, if any failures occur during the workflow(i.e., failure of build rv32emu executable, segmentation fault, ...),  then the workflow will send an email to notify maintainer or committer. Of course, no new benchmark data will be updated in the web visualization of benchmark.

Originally, I want to comment the custom message in the thread of PR when unexpected condition happens, but it seems like this feature has not been implemented yet, see this [issue](https://github.com/benchmark-action/github-action-benchmark/issues/172)

qwe661234 commented:
> > What information we would receive if the benchmark fails?
> 
> Which part of failure in workflow that you referred to?
> 
> Normally, if any failures occur during the workflow(i.e., failure of build rv32emu executable, segmentation fault, ...), then the workflow will send an email to notify maintainer or committer. Of course, no new benchmark data will be updated in the web visualization of benchmark.
> 
> Originally, I want to comment the custom message in the thread of PR when unexpected condition happens, but it seems like this feature has not been implemented yet, see this [issue](https://github.com/benchmark-action/github-action-benchmark/issues/172)

Some errors only occur when running coreMark (e.g. JIT mode with clang compiler), so other workflows will pass and won't send emails to notify the maintainer or committer. This is the reason why I'm asking you this question, but it is not a critical problem.

jserv commented:
Thank @ChinYikMing for contributing!

jserv commented:
> After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.
> Last, the benchmark CI should work!

@ChinYikMing, Please check the [CI results](https://github.com/sysprog21/rv32emu/actions/runs/6146294738) and add the hyperlink to benchmark visualization as a pull request for `README.md`.
